Data-Structure/Algorithm-Design-Part-1/DictionaryUsingBST/BSTToImplementDictionary.java

```diff
@@ -0,0 +1,241 @@
+package DictionaryUsingBST;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.json.simple.JSONObject;
+
+/**
+ * @author Niraj Gupta
+ * Class to implement the Dictionary interface using Binary search tree.
+ * 
+ * @param <K> : The Data Type assigned to the key of the node.
+ * @param <V> : The Data Type assigned to the value of the node.
+ */
+public class BSTToImplementDictionary<K extends Comparable<K>, V> implements Dictionary<K, V>{
+	BSTNode<K, V> root;
+	JSONObject jsonObject;
+	private List<BSTNode<K, V>> sortedList = new ArrayList<BSTNode<K, V>>();
+
+	/**
+	 * Parameterized Constructor
+	 * @param jsonObject : The parsed JSONObject.
+	 */
+	public BSTToImplementDictionary(JSONObject jsonObject){
+		this.jsonObject = jsonObject;
+	}
+
+	/**
+	 * Non-Parameterized Constructor.
+	 */
+	public BSTToImplementDictionary(){
+		this.root = null;
+	}
+
+	/**
+	 * Utility Method to insert data into the tree by calling the insert method.
+	 * @param jsonObject : The parsed JSONObject containing the key-value pair 
+	 * 						which is to be inserted into the tree.
+	 * @throws DuplicateElementInsertionException : When the node to be inserted in the tree 
+	 * 												already exists with same key in the tree. 
+	 */
+	public void insertData(JSONObject jsonObject) throws DuplicateElementInsertionException{
+		for(Object key : jsonObject.keySet()){
+			insert((K) key, (V) jsonObject.get(key));
+		}
+	}
+
+	public void insert(K key, V value) throws DuplicateElementInsertionException{
+		if(key != null && value != null){
+			if(root == null){
+				root = new BSTNode<K, V>(key, value);
+			}else{
+				insertRecursively(root, key, value);
+			}
+		}else{
+			throw new TheKeyValuePairIsNullException("Invalid Input : The key value pair is null.");
+		}
+	}
+
+	/**
+	 * Utility method to insert node into the tree recursively.
+	 * @param node : The node is inserted into the tree recursively.
+	 * @param key : The node which uniquely identify the value.
+	 * @param value : The value which is stores in the node corresponding to a key.
+	 * @throws DuplicateElementInsertionException : When the duplicate node is inserted into the tree.
+	 */
+	private void insertRecursively(BSTNode<K, V> node, K key, V value) throws DuplicateElementInsertionException{
+		if(key.compareTo(node.getKey()) == 0){

```

Comment:
good validations but u missed out on null check here :)

Reply:
More improvement is required. Will study more about null pointer exception.